rfcs/0056-FE-APIs.md

```diff
@@ -0,0 +1,234 @@
+        content: <ReviewWorkflowsSidePanel />,
+      },
+    ]);
+
+    app.addDocumentHeaderAction([
+      {
+        label: "Select Locale",
+        options: [{ label: "english", value: "en" }],
+      },
+    ]);
+  },
+};
+
+// admin/src/documentActions.ts
+export const PublishMultipleLocales = {
+  label: "Publish multiple locales",
+  onClick: () => window.alert("you're publishing!"),
+};
+```
+
+# Motivation
+
+Current injectionZones let you inject components into areas of the content-manager, while this is a good it's a somewhat limiting API, the `addDocumentHeaderAction` & `addEditViewSidePanels` APIs while the former may seem very constricted provides a balance between user's being able to add actions quicker but not affect the design of the product, the latter is incredibly flexible but a much clearer typed system compared to the current implementation of `InjectionZones`. The `addDocumentActions` API is a new API that allows you to add actions to be rendered in the DocumentActions menu (this is new to V5).
+
+Overall, we want to move towards more developer friendly APIs that are quicker to use, more intuitive yet do not cause concerns for the usability of our product.
+
+# Detailed proposal
+
+> ⚠️ All types are rough and have not been validated, there may be reason to loosen types based on design specs.
+
+We’ll add the new APIs to the `app` class to keep with the current imperative model of introducing things to the Strapi admin panel. These APIs will be overloaded so you can apply the addition so it’s always rendered, or you can conditionally apply it based on the context provided.
+
+All the APIs dedicated to the EditView of the content manager will receive similar context payloads should user’s opt in to use the function overload, it can be illustrated like this:
+
+```tsx
+interface EditViewContext {
+  activeTab: "draft" | "published"; // room for more.
+  id: string;
+  meta: DocumentMeta;
+  document: Document;
+}
+
+// TBD – placeholder type
+interface DocumentMeta {
+  publishedAt: string;
+  createdAt: string;
+  createdBy: string;
+  // other dimensions e.g. locale
+  [key: string]: any;
+}
+
+// TBD – placeholder type
+interface Document {
+  // User attributes
+  [key: string]: any;
+}
+```
+
+We can then apply it to the APIs in a uniform fashion which creates consistency:
+
+```tsx
+interface ContentManagerEditViewAPIs {
+  funcName<TConfig extends object>(config: TConfig[]): void;
+  funcName<TConfig extends object>(configFn: ConfigFn<TConfig>): void;
+}
+
+type ConfigFn<TConfig extends object> = (
+  prevConfigs: TConfig[],
+  ctx: EditViewContext
+) => TConfig[];
+```
+
+Because the function receives the previous list of configs, we expect the complete list of configs to be returned, this means if a plugin required it could remove a configuration, or mutate a current configuration building upon one another e.g. the publish action would be able to be mutated by the `i18n` plugin correctly.
```

Comment:
I see 1 issue with this implementation if there are no controls for the order the configs are read in how can I chance something that i18n changed if it is loaded after my function is done.